### PR TITLE
Narrow from 'any' with all type guards

### LINKED ISSUE #4432: User-defined type guards don't work for all types
If the user defines a type guard for a boolean, as in this example:

```TypeScript
function isBool(a: any): a is boolean {
    return (typeof a === "boolean");
}

function bob() {
    var z;

    if (isBool(z)) {
        z; // type of z is still 'any'
    }
}
```

...then z is not properly refined to 'boolean'.  It stays 'any'.  I think the expected behavior here is for the type guard to refine to boolean inside of the if check.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #4432.

### CODE REVIEW

**Comment by tinganho** [comment]
I was about to submit the same PR yesterday. https://github.com/tinganho/TypeScript/commit/3ea297d3b34bc4dfd62c9f74b4f3004a5f663c19

**Comment by DanielRosenwasser** [comment]
Sorry you went through that, I didn't mean to step on anyone's feet! Though in the future if you'd like to take an assigned bug on, just post a quick message on the issue saying so.

**Comment by tinganho** [comment]
Oh no problem. I was just thinking it was a small change and that I could quickly submit it. I didn't lose that much time. I will send a small notification next time.

File: tests/baselines/reference/typeGuardsWithInstanceOfByConstructorSignature.errors.txt [hunk 1]
```diff
@@ -31,6 +41,8 @@ tests/cases/conformance/expressions/typeGuards/typeGuardsWithInstanceOfByConstru
     if (obj2 instanceof A) { // can't narrow type from 'any'
```
Comment (mhegazy) [inline tests/baselines/reference/typeGuardsWithInstanceOfByConstructorSignature.errors.txt]: comment out dated. please update.

File: tests/baselines/reference/typeGuardsWithInstanceOfByConstructorSignature.errors.txt [hunk 2]
```diff
@@ -91,7 +105,11 @@ tests/cases/conformance/expressions/typeGuards/typeGuardsWithInstanceOfByConstru
     if (obj6 instanceof C) { // can't narrow type from 'any'
```
Comment (mhegazy) [inline tests/baselines/reference/typeGuardsWithInstanceOfByConstructorSignature.errors.txt]: same here..

File: tests/baselines/reference/typeGuardsWithInstanceOfByConstructorSignature.errors.txt [hunk 3]
```diff
@@ -112,6 +130,8 @@ tests/cases/conformance/expressions/typeGuards/typeGuardsWithInstanceOfByConstru
     if (obj8 instanceof D) { // can't narrow type from 'any'
```
Comment (mhegazy) [inline tests/baselines/reference/typeGuardsWithInstanceOfByConstructorSignature.errors.txt]: and all other similar comments in the file :)

**Comment by mhegazy** [comment]
LGTM, a few comments though
- please update test comments
- please run the RWC tests and verify if there are any unexpected failures/ failures we need to know about
- log a spec issue to update the spec.

**Comment by mhegazy** [comment]
We should close this one as well.

**Comment by vvakame** [comment]
Is this commit already merged in master? I :heart: this changes.

**Comment by DanielRosenwasser** [comment]
@vvakame, no, the problem is that `instanceof` type guards can have bad behavior when you do something like `x instanceof Object` - now you can access effectively nothing. By extension, user-defined type guards are weakened down to `instanceof` guards. It's not optimal, but we're still considering the issue.

**Comment by vvakame** [comment]
I understand!

**Pull request closed without merging**